uaclient/tests/test_gpg.py

```diff
@@ -0,0 +1,70 @@
+from uaclient import exceptions, gpg, util
+from uaclient.testing import data
+
+import logging
+import os
+import pytest
+
+
+class TestExportGPGKeyFromKeyring:
+
+    def test_key_error_on_missing_key(self, tmpdir):
+        """Raise UserFacingError when key_id is not in source_keyring_file."""
+        src_keyring = tmpdir.join('ubuntu-advantage-keyring')
+        src_keyring.write('')
+        destination_keyfile = tmpdir.join('destination_keyfile').strpath
+        key_id = 'Bueller... Bueller'
+        with pytest.raises(exceptions.UserFacingError) as excinfo:
+            gpg.export_gpg_key_from_keyring(
+                key_id=key_id, source_keyring_file=src_keyring.strpath,
+                destination_keyfile=destination_keyfile)
+
+        error_msg = "GPG key '%s' not found in %s" % (
+            key_id, src_keyring.strpath)
+        assert error_msg in str(excinfo.value)
+        assert False is os.path.exists(destination_keyfile)
+
+    @pytest.mark.parametrize('caplog_text', [logging.ERROR], indirect=True)
+    def test_user_facing_error_on_invalid_keyring_packet(
+            self, caplog_text, tmpdir):
+        """Raise UserFacingError on invalid keyring packet. Log cmd error."""
+        source_keyring = tmpdir.join('keyring').strpath
+        destination_keyfile = tmpdir.join('destination_key').strpath
+        key_id = '3CB3DF682220A643B43065E9B30EDAA63D8F61D0'
+        # Create bogus keyring with invalid packet which causes gpg errors
+        with open(source_keyring, 'wb') as stream:
+            stream.write(b'\n')
```

Comment:
Not worth changing, but for reference I _think_ we could do this with `tmpdir.join('keyring').write(b'\n', mode='wb')`.

Reply:
I'll change it here for reference next time. I'd rather use the builtin features of pytest tmpdir anyway.